bin/gftools-push-status.py

```diff
@@ -0,0 +1,181 @@
+states.
+
+
+Usage:
+
+Check push status for the past month:
+gftools push-status path/to/google/fonts
+
+Check push status for specified date:
+gftools push-status path/to/google/fonts -pd 2020-07-01
+"""
+import argparse
+from argparse import RawDescriptionHelpFormatter
+import os
+from pathlib import Path
+import gftools.fonts_public_pb2 as fonts_pb2
+from google.protobuf import text_format
+from datetime import datetime, timedelta
+import json
+import requests
+
+
+DEV_META_URL = "http://fonts-dev.sandbox.google.com/metadata/fonts"
+SANDBOX_META_URL = "http://fonts.sandbox.google.com/metadata/fonts"
+PROD_META_URL = "http://fonts.google.com/metadata/fonts"
+
+
+def get_family_metadata(url):
+    """Get family json data from a Google Fonts metadata url"""
+    # can't do requests.get("url").json() since request text starts with ")]}'"
+    info = json.loads(requests.get(url).text[5:])
+    return {i['family']: i for i in info["familyMetadataList"]}
+
+
+DEV_META = get_family_metadata(DEV_META_URL)
+SANDBOX_META = get_family_metadata(SANDBOX_META_URL)
+PROD_META = get_family_metadata(PROD_META_URL)
+
+# If no push date is added we can assume that the push has probably
+# happened in the past month
+ONE_MONTH_AGO = datetime.now() - timedelta(days=31)
+
+
+def families_from_file(fp):
+    """Convert to_sandbox.txt and to_production.txt files to a list of
+    family names."""
+    results = set()
+    with open(fp, "r") as doc:
+        family_dirs = doc.read().split()
+
+    gf_path = Path(fp).parent
+    for dir_ in family_dirs:
+        metadata_file = gf_path.joinpath(f"{dir_}/METADATA.pb")
+        if not metadata_file.is_file():
+            raise FileNotFoundError(f"No METADATA.pb file found in {dir_}")
+        metadata = fonts_pb2.FamilyProto()
+        with open(metadata_file, 'rb') as meta:
+            text_format.Parse(meta.read(), metadata)
+            results.add(metadata.name)
+    return results
+
+
+def families_status(info, push_date, filter_families=set()):
+    """Determine which families have been pushed or not pushed within an
+    allocated date. If a filter_families set is provided, remove families
+    from the results which are not included in this set."""
+    results = {"pushed": set(), "not_pushed": set()}
+    if filter_families:
+        info = {k: v for k,v in info.items() if k in filter_families}
+    for family in info:
+        last_push = info[family]['lastModified']
+        last_push = datetime.strptime(last_push, "%Y-%m-%d")
+        if last_push > push_date:
+            results['pushed'].add(family)
+        if last_push < push_date:
+            results["not_pushed"].add(family)
+    return results
+
+
+def status_reporter(server_name, status):
+    """Produce a report for a server status dict"""
+    report = []
+    report.append(f"***{server_name}***")
+    pushed = list(sorted(status["pushed"]))
+    not_pushed = list(sorted(status["not_pushed"]))
```

Comment:
If you make a function that gives the status then you can group by status then loop over the possible status values and shove things into the report rather than having to write if/else and have variables for each status